onnxruntime/test/providers/kernel_def_hash_test.cc

```diff
@@ -0,0 +1,171 @@
+
+using json = nlohmann::json;
+
+namespace onnxruntime {
+namespace test {
+
+namespace {
+// if set to 1, treat skipping of the kernel def hash tests as a test failure
+static constexpr const char* kRunKernelDefHashTestOrFailEnvVar =
+    "ORT_TEST_RUN_KERNEL_DEF_HASH_TEST_OR_FAIL";
+
+std::string DumpKernelDefHashes(const onnxruntime::KernelDefHashes& kernel_def_hashes) {
+  const json j(kernel_def_hashes);
+  return j.dump(/* indent */ 4);
+}
+
+KernelDefHashes ParseKernelDefHashes(std::istream& in) {
+  KernelDefHashes kernel_def_hashes{};
+  const json j = json::parse(in);
+  j.get_to<onnxruntime::KernelDefHashes>(kernel_def_hashes);
+  return kernel_def_hashes;
+}
+
+KernelDefHashes ReadKernelDefHashesFromFile(const PathString& path) {
+  std::ifstream in{path};
+  ORT_ENFORCE(in, "Failed to open file: ", ToMBString(path));
+  const auto kernel_def_hashes = ParseKernelDefHashes(in);
+  return kernel_def_hashes;
+}
+
+void CheckKernelDefHashes(const KernelDefHashes& actual, const KernelDefHashes& expected) {
+  ASSERT_TRUE(std::is_sorted(actual.begin(), actual.end()));
+  ASSERT_TRUE(std::is_sorted(expected.begin(), expected.end()));
+
+  constexpr const char* kNoteReference = "Note: Please read the note at the top of this file: " __FILE__;
+
+  KernelDefHashes expected_minus_actual{};
+  std::set_difference(expected.begin(), expected.end(), actual.begin(), actual.end(),
+                      std::back_inserter(expected_minus_actual));
+  EXPECT_TRUE(expected_minus_actual.empty())
+      << "Some expected kernel def hashes were not found.\n"
+      << kNoteReference << "\n"
+      << DumpKernelDefHashes(expected_minus_actual);
+
+  KernelDefHashes actual_minus_expected{};
+  std::set_difference(actual.begin(), actual.end(), expected.begin(), expected.end(),
+                      std::back_inserter(actual_minus_expected));
+  EXPECT_TRUE(actual_minus_expected.empty())
+      << "Unexpected kernel def hashes were found, please update the expected values as needed "
+         "(see the output below).\n"
+      << kNoteReference << "\n"
+      << DumpKernelDefHashes(actual_minus_expected);
+}
+}  // namespace
+
+TEST(KernelDefHashTest, DISABLED_PrintCpuKernelDefHashes) {
+  KernelRegistry kernel_registry{};
+  ASSERT_STATUS_OK(RegisterCPUKernels(kernel_registry));
+  const auto cpu_kernel_def_hashes = kernel_registry.ExportKernelDefHashes();
+  std::cout << DumpKernelDefHashes(cpu_kernel_def_hashes) << "\n";
+}
+
+TEST(KernelDefHashTest, ExpectedCpuKernelDefHashes) {
+  // this test should only run in a build containing exactly the set of CPU
+  // kernels that can be used in ORT format models
+  const bool is_enabled = []() {
+#if !defined(DISABLE_CONTRIB_OPS) &&       \
+    !defined(DISABLE_ML_OPS) &&            \
+    !defined(ML_FEATURIZERS) &&            \
+    !defined(BUILD_MS_EXPERIMENTAL_OPS) && \
+    !defined(ENABLE_TRAINING) && defined(ENABLE_TRAINING_OPS)
+    return MlasNchwcGetBlockSize() > 1;
+#else
```

Comment:
comment saying NCHWc kernels are disabled if MlasNchwcGetBlockSize() <= 1 would be helpful as that's a somewhat cryptic connection. 

may also need a note saying we don't expect NCHWc kernels to be used with a minimal build. this could however change if we have x64 edge devices running a minimal build in the future.